feat(sort) make orderBy work with locality by using Intl.collator | give…

wickstargazer opened the pull request:
… an option for custom compare if def[0] is a function

Fixes #562 (it's a good idea to open an issue first for discussion)

- [x] - `npm test` succeeds
- [ ] - Code coverage does not decrease (if any source code was changed)
- [ ] - If applicable, appropriate JSDoc comments were updated in source code (if applicable)
- [ ] - If applicable, approprate changes to js-data.io docs have been suggested ("Suggest Edits" button)

crobinson42 commented on src/Query.js:
```diff
@@ -263,13 +263,21 @@ export default Component.extend({
+    if (utils.isNumber(cA) || utils.isNumber(cB)) {
+      isNumeric = true
+    }
+    const collator = new Intl.Collator(locale, {
```
Will this throw if `Intl` is not defined in an environment?

wickstargazer replied:
Ok i will check for undefined and add it the default behavior, that would help?

crobinson42 replied:
Yeah, definitely need that. This won't run in Node.js env at all

zuzusik replied:
@crobinson42 looks like `Intl` API is available in `Node.js` after all - I've checked it starting from version `8.11.1` (oldest one on my machine) and it works just fine. I am using `Node` installed by `nvm` on `macOS`.

crobinson42 replied:
I suppose it wouldn't hurt to only support Node >= 8 in v4 release.. I notice a lot of popular npm packages are dropping support for old Node.js versions. What do you guys think?

crobinson42 replied:
We will still need either a polyfill or a graceful failover for `Intl` api: https://caniuse.com/#search=Intl

zuzusik replied:
I think it's a good idea to drop Node below 8.

As for `Intl`, actually `Collator.compare` part of this API is pretty much widely supported: https://caniuse.com/#feat=mdn-javascript_builtins_intl_collator_compare

I wouldn't say we need polyfill, unless we want to support IE 10 and below (which I would drop as well)

wickstargazer replied:
I agree that there is no need to support IE 10 if newer browser works, i already added support for edge by force setting locale...
Other than that i can add to the docs and help write a test for the #540 and call it a new day?

wickstargazer replied:
I agree that there is no need to support IE 10 if newer browser works, i already added support for edge by force setting locale...
Other than that i can add to the docs and help write a test for the #540 and call it a new day?

crobinson42 replied:
@wick Sounds good 👍

wickstargazer replied:
updated the doc, not sure where to put the available since version :)

wickstargazer commented:
Or is it better to add this Implementation 

https://www.npmjs.com/package/intl-collator

crobinson42 commented:
If we can avoid adding new pkg dependencies that is preferable.


> On Jan 31, 2020, at 9:38 AM, Wick <notifications@github.com> wrote:
> 
> Or is it better to add this Implementation
> 
> https://www.npmjs.com/package/intl-collator <https://www.npmjs.com/package/intl-collator>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/js-data/js-data/pull/563?email_source=notifications&email_token=ABNSMS57E3IPWJTSXW4J6FDRAROYXA5CNFSM4KN3WS4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKPNAQI#issuecomment-580833345>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABNSMSZSFE3MTJXFKRK5VTLRAROYXANCNFSM4KN3WS4A>.
>

wickstargazer commented:
Ok then i would add to the docs incase anyone wants to use the feature install extra package in nodejs environment:)

crobinson42 commented:
https://www.js-data.io/v4.0 suggest edits. I would say they belong under queries around `orderBy` syntax

wickstargazer commented:
I was in the sense of it being supported in node as well when googling but havent tested it out, thankz @zuzusik :) i will implement a default failover and we can add the support into the documentation i see safari still wont support it?

zuzusik commented:
@crobinson42 good call to check https://github.com/js-data/js-data/issues/540 with this one 👍  - it turns out #540 actually gets fixed by this - `Collator.compare` uses same logic as `Array.sort` regarding `null` and `undefined`, putting them at the very end of array in `asc` mode, which is exactly what required in #540.

I will just add appropriate tests tomorrow - will open a PR to `wickstargazer:feature/sort-by-locality`

zuzusik commented:
@wickstargazer btw, I think this should be pointing to `v4` branch, not `master`

zuzusik commented:
@crobinson42 what do you think - should this be released as `3.x` and then ported to `v4` or should it go to `v4` directly?

zuzusik commented:
@wickstargazer I have opened a PR in your fork to add tests for #540 https://github.com/wickstargazer/js-data/pull/1

wickstargazer commented:
would love for it to be in 3.x as its running on our production and 4 is still a distance away...will also merge it into 4

crobinson42 commented:
We can't release in v3 unless there is fallback for non supporting environments because that's a breaking change. v3 supports ie11 and node 0.12 and up... So, if we're going to release for v3 then that requires graceful failing.

wickstargazer commented:
Hey guys, not an issue if want to put it for v4 only. Or even if i should have a fallback for v3 .. I will be doing some stuff tonight .. 
@zuzusik  the default is "en" for the timebeing IF the browser is Edge. We tested on production, it does not break on other browsers except Edge if locale is null.

wickstargazer commented:
Ok guys, i have changed it to be from v4 and optimized the code a little, also add locale to the reserved keyword (which i am not sure should i do)

@zuzusik  for the Int.Collator i added a default fallback to 'en' Although its mentioned here 
[Intl.Collator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl#Locale_negotiation)

> the "best fit" matcher lets the runtime provide a locale that's at least, but possibly more, suited for the request than the result of the Lookup algorithm. If the application doesn't provide a locales argument, or the runtime doesn't have a locale that matches the request, then the runtime's default locale is used

So i am guessing although default is provided and locale is not passed it, the sort should work by locality anyways.

crobinson42 commented:
@wickstargazer 201 file changes?

wickstargazer commented:
i think its because i rebase the forked branch to the original v4 and push force
maybe i should re-merge, let me do it from another branch ....

crobinson42 commented:
@wickstargazer Thanks for this effort! You mentioned you _need_ this in v3 - if you want to PR this feature to v3 and make it optional I'll work with you to get a release out 👍🏼

wickstargazer commented:
i think we can live until the v4 comes out, we are going to do a small restrcuture in april .. can wait till then!
